### PR TITLE
Support UDS & TCP/IP for ipc

### LINKED ISSUE #68: Move from TCP/IP Sockets to Unix Domain Sockets
We're currently working with Servers that use TCP/IP as protocol.

Since we had in the past everything connected through UDS, and in the Go side (AFAIK @autholykos ) they're still using them, we should migrate away from TCP/IP to UDS for the tests and the default host/port usage.

### LINKED ISSUE #69: Handle better addresses and ports used for testing in Travis and locally
We currently have the following problem:

When we work locally and run rusk tests, we need to set a `host` and `port` which by default are: `127.0.0.1` (loopback addr) and `8085` respectively.

The problem is that inside of the Travis build, `127.0.0.1` has firewall rules applied and the loopback address used for testing is `0.0.0.0`.

Therefore, we need to constantly change the values depending on the environment we're working on. Therefore we should try to find a solution to this problem at some point.

Note also that as soon as #68 get's addressed, we will need to figure out how to do it for UDS.

### PR DESCRIPTION
**Pull request by CPerezz**
Since rusk will ommunicate with the Golang node and this process can happen on Windows and Unix-based machines, we need to provide UDS for the second ones and TCP/IP for all of them when working with GRPC communications.

- Added tests to generate a client and a server that work through UDS.
- Refactored the bin app to startup using UDS toghether with .
- Added an Arg option to set a custom path for UDS filedescriptor bind.

The solution is a bit messy just for the `unix.rs` mod that I need to define twice. But I don't think we should export it to be able to use it on the tests. That's why I duplicated just this code.

For the rest, it should be OK.

Closes #68 
Closes #69 IMO since we now don't have any address problems with the Travis or local testing at the moment wit the latest addresses we were using.

### CODE REVIEW

**Comment by CPerezz** [comment]
# UPDATE

We need to also support Windows machines. Therefore, we will keep the previous TCP/IP impl for Windows systems and use UDS for the other ones.

We will allow to select one of the two methods via CLI arg

File: Cargo.toml [hunk 1]
```diff
@@ -22,9 +22,11 @@ clap = "2.33.3"
 prost = "0.6"
 kelvin = "0.18"
 kelvin-radix = "0.10"
+futures = { version = "0.3", default-features = false, features = ["alloc"] }
 thiserror = "1.0"
 anyhow = "1.0"
 rustc_tools_util = "0.2"
+tower = "0.3"
```
Comment (vlopes11) [inline Cargo.toml:29]: I heard somewhere tower is discontinued. Are we sure that isn't the case?

**Review by vlopes11 — APPROVED** [review]
LGTM, lets just reassure tower is still active

File: Cargo.toml (hunk 1 above)
Comment (CPerezz) [reply Cargo.toml:29]: It's needed by the tests.

See:
```
// Create the client binded to the default testing UDS path.
        let channel = Endpoint::try_from("http://[::]:50051")?
            .connect_with_connector(service_fn(|_: Uri| {
                // Connect to a Uds socket
                UnixStream::connect(SOCKET_PATH)
            }))
            .await?;
        let mut client = EchoerClient::new(channel);
```
Tower provides us this `service_fn(|_|...` which we need to setup the `Channel`. If you know any workarround, tell me.

File: src/bin/main.rs [hunk 2]
```diff
@@ -22,11 +29,29 @@ async fn main() {
...
+                .value_name("socket")
+                .help("Path for setting up the UDS ")
+                .default_value(SOCKET_PATH)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("ipc_method")
+                .long("ipc_method")
+                .value_name("ipc_method")
+                .possible_values(&["uds", "tcp_ip"])
```
Comment (vlopes11) [inline src/bin/main.rs:45]: `uds` isn't possible with windows. Maybe the provided `possible_values` will change depending on the target os?

File: src/bin/main.rs [hunk 3]
```diff
@@ -70,19 +95,68 @@ async fn main() {
...
+        Some(method) => match (cfg!(windows), method) {
+            (_, "tcp_ip") => {
+                startup_with_tcp_ip(
+                    matches.value_of("host").unwrap_or_else(|| HOST_ADDRESS),
+                    matches.value_of("port").unwrap_or_else(|| PORT),
+                )
+                .await
+            }
+            (true, "uds") => {
+                panic!("Windows does not support Unix Domain Sockets");
```
Comment (vlopes11) [inline src/bin/main.rs:110]: Should be `unreachable` instead of `panic` in case the previous comment is fixed

File: src/bin/main.rs [hunk 4]
```diff
@@ -70,19 +95,68 @@ async fn main() {
...
-        matches.value_of("port").unwrap_or_else(|| PORT),
-    )
-    .await
-    {
-        Ok(_) => (),
+    // Match the desired IPC method. Or set the default one depending on the OS used.
+    // Then startup rusk with the final values.
+    let res = match matches.value_of("ipc_method") {
+        Some(method) => match (cfg!(windows), method) {
+            (_, "tcp_ip") => {
```
Comment (vlopes11) [inline src/bin/main.rs:102]: Is it critical we add ssl capability now?

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: src/bin/main.rs (hunk 4 above)
Comment (CPerezz) [reply src/bin/main.rs:102]: I don't think we should overcomplicate things now. We should go for a MVP that allows us to implement the services needed by rusk.

We always have time to add more things later on.

File: src/bin/main.rs (hunk 2 above)
Comment (CPerezz) [reply src/bin/main.rs:45]: I tried to make a cgf_attrib. But makes the code less readable.
Therefore, I accept on the arg the two options and parse later depending on the `target_os` the options choosen.

File: src/bin/main.rs (hunk 3 above)
Comment (CPerezz) [reply src/bin/main.rs:110]: Probably not going to modify the previous unless it's something you consider critical.

**Review by vlopes11 — APPROVED** [review]
LGTM

**Pull request merged**